addons/website_sale_picking/models/sale_order.py

```diff
@@ -0,0 +1,23 @@
+from odoo import models
+
+
+class SaleOrder(models.Model):
+    _inherit = "sale.order"
+
+    def _check_carrier_quotation(self, force_carrier_id=None, keep_carrier=False):
+        res = super()._check_carrier_quotation(
+            force_carrier_id=force_carrier_id,
+            keep_carrier=keep_carrier,
+        )
+        if res:
+            fpos_before = self.fiscal_position_id
+            if self.carrier_id.delivery_type == 'onsite' and self.carrier_id.warehouse_id:
+                self.partner_shipping_id = self.carrier_id.warehouse_id.partner_id
+            else:
+                # setting partner_shipping_id as the carrier pickup location
+                # overrides the original partner shipping address
+                # so it needs to be recomputed if the delivery method is not onsite picking
+                self._compute_partner_shipping_id()
```

Comment:
Only recompute the partner shipping if the carrier was onsite before the super call ?
Because here, you will reset the chosen shipping address of customers.